device/coordinate_manager.cpp

```diff
@@ -0,0 +1,184 @@
+                throw std::runtime_error("Duplicate physical y coordinate found in the worker cores");
+            }
+            physical_y_to_logical_y[*physical_y_it] = logical_y;
+            logical_y++;
+            physical_y_it++;
+        } else {
+            physical_y_it++;
+        }
+    }
+
+    auto physical_x_it = physical_x_unharvested.begin();
+    std::size_t logical_x = 0;
+    for(std::size_t x = 0; x < worker_grid_size.x; x++) {
+        if (x_to_harvest.find(x) == x_to_harvest.end()) {
+            logical_x_to_physical_x[logical_x] = *physical_x_it;
+            if (physical_x_to_logical_x.find(*physical_x_it) != physical_x_to_logical_x.end()) {
+                throw std::runtime_error("Duplicate physical x coordinate found in the worker cores");
+            }
+            physical_x_to_logical_x[*physical_x_it] = logical_x;
+            logical_x++;
+            physical_x_it++;
+        } else {
+            physical_x_it++;
+        }
+    }
+}
+
+void CoordinateManager::fill_logical_to_virtual_mapping(const std::set<size_t>& physical_x_unharvested, const std::set<size_t>& physical_y_unharvested) {
+    auto physical_y_it = physical_y_unharvested.begin();
+    for (std::size_t y = 0; y < logical_y_to_virtual_y.size(); y++) {
+        logical_y_to_virtual_y[y] = *physical_y_it;
+        if (virtual_y_to_logical_y.find(*physical_y_it) != virtual_y_to_logical_y.end()) {
+            throw std::runtime_error("Duplicate virtual y coordinate found in the worker cores");
+        }
+        virtual_y_to_logical_y[*physical_y_it] = y;
+        physical_y_it++;
+    }
+
+    auto physical_x_it = physical_x_unharvested.begin();
+    for (std::size_t x = 0; x < logical_x_to_virtual_x.size(); x++) {
+        logical_x_to_virtual_x[x] = *physical_x_it;
+        if (virtual_x_to_logical_x.find(*physical_x_it) != virtual_x_to_logical_x.end()) {
+            throw std::runtime_error("Duplicate virtual x coordinate found in the worker cores");
+        }
+        virtual_x_to_logical_x[*physical_x_it] = x;
+        physical_x_it++;
+    }
+}
+
+#include "device/blackhole/blackhole_coordinate_manager.h"
+#include "device/grayskull/grayskull_coordinate_manager.h"
+#include "device/wormhole/wormhole_coordinate_manager.h"
+
+CoordinateManager* CoordinateManager::get_coordinate_manager(
+    tt::ARCH arch,
+    const tt_xy_pair& worker_grid_size,
+    const std::vector<tt_xy_pair>& workers,
+    std::size_t harvesting_mask) {
+
+    if (arch == tt::ARCH::GRAYSKULL) {
+        return new GrayskullCoordinateManager(worker_grid_size, workers, harvesting_mask);
+    } else if (arch == tt::ARCH::WORMHOLE_B0) {
+        return new WormholeCoordinateManager(worker_grid_size, workers, harvesting_mask);
+    } else if (arch == tt::ARCH::BLACKHOLE) {
+        return new BlackholeCoordinateManager(worker_grid_size, workers, harvesting_mask);
+    } else {
+        throw std::runtime_error("Invalid architecture for coordinate manager");
+    }
+}
```

Comment:
Consider returning a `std::unique_ptr<CoordinateManager>`?

Reply:
Also
```cpp
switch (arch) {
    case tt::ARCH::GRAYSKULL:
        return ...
    ...
}

throw ...
```
might be less error-prone. I believe there's a clang-tidy rule to check that all enumerated cases are handled if the enum ever grows.